pkg/registry/apis/provisioning/jobs/syncer.go

```diff
@@ -1,469 +1,447 @@
-	var version string
-	switch item.Resource {
-	case folders.RESOURCE:
-		version = folders.VERSION
-	case dashboards.DASHBOARD_RESOURCE:
-		version = dashboardsv2alpha1.VERSION
-	default:
-		return fmt.Errorf("unknown resource api version: %s", item.Resource)
-	}
-
-	client := r.client.Resource(schema.GroupVersionResource{
-		Group:    item.Group,
-		Version:  version,
-		Resource: item.Resource,
-	})
-
-	if err := client.Delete(ctx, item.Name, metav1.DeleteOptions{}); err != nil {
-		return fmt.Errorf("delete resource: %w", err)
 	}
-
-	return nil
 }
 
-// replicateTree replicates all files in the repository.
-func (r *Syncer) replicateTree(ctx context.Context, ref string) error {
-	// Load the tree of the repository
-	fileTree, err := r.repository.ReadTree(ctx, ref)
+// Convert git changes into resource file changes
+func (r *syncJob) applyVersiondChanges(ctx context.Context, repo repository.VersionedRepository, previousRef, currentRef string) error {
+	diff, err := repo.CompareFiles(ctx, previousRef, currentRef)
 	if err != nil {
-		return fmt.Errorf("read tree: %w", err)
+		return fmt.Errorf("compare files error: %w", err)
 	}
 
-	// Load the list of resources in the cluster
-	list, err := r.lister.List(ctx, r.client.GetNamespace(), r.repository.Config().Name)
-	if err != nil {
-		return fmt.Errorf("list resources: %w", err)
-	}
-
-	// Remove resources not longer present in the repository
-	if err := r.cleanUnnecessaryResources(ctx, fileTree, list); err != nil {
-		return fmt.Errorf("clean up before replicating tree: %w", err)
-	}
-
-	// Create folders
-	folderTree := resources.NewFolderTreeFromResourceList(list)
-	for _, entry := range fileTree {
-		parent := resources.ParentFolder(path.Dir(entry.Path), r.repository.Config())
-		if err := r.ensureFolderPathExists(ctx, path.Dir(entry.Path), parent, folderTree); err != nil {
-			return fmt.Errorf("ensure folder path exists: %w", err)
-		}
-	}
-
-	// Create a list of hashes to identify changed files
-	hashes := make(map[string]string)
-	for _, item := range list.Items {
-		hashes[item.Path] = item.Hash
+	if len(diff) < 1 {
+		message := "no changes detected between commits"
+		r.syncStatus.State = provisioning.JobStateSuccess
+		r.syncStatus.Message = append(r.syncStatus.Message, message)
+		r.syncStatus.Incremental = true
+		r.jobStatus.Message = message
+		return nil
 	}
 
-	// Create Dashboards
-	for _, entry := range fileTree {
-		logger := logging.FromContext(ctx).With("file", entry.Path)
-		if hashes[entry.Path] == entry.Hash {
-			logger.Debug("file is up to date")
-			continue
-		}
-
-		if !entry.Blob {
-			logger.Debug("ignoring non-blob entry")
-			continue
-		}
-
-		if resources.ShouldIgnorePath(entry.Path) {
-			logger.Debug("ignoring file")
-			continue
-		}
-
-		info, err := r.repository.Read(ctx, entry.Path, ref)
-		if err != nil {
-			return fmt.Errorf("read file: %w", err)
+	for _, change := range diff {
+		if len(r.jobStatus.Errors) > 20 {
```

Comment:
It's weird how we are handling errors vs the job status. I would have expected this to return an error to the caller.

Reply:
depends if we want to fail or continue -- this changes our logic so that *when* something fails we report it in the job status rather than aborting the rest of the sync.

I am 💯 sure this structure and our error handling/reporting will need some work!